File: Inxi.NET/Hardware/HardDrive.vb

'    Inxi.NET  Copyright (C) 2020-2021  EoflaOE
'
'    This file is part of Inxi.NET
'
'    Inxi.NET is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Inxi.NET is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Public Class HardDrive
    Inherits HardwareBase

    ''' <summary>
    ''' The name of hard drive
    ''' </summary>
    Public Overrides ReadOnly Property Name As String
    ''' <summary>
    ''' The udev ID of hard drive
    ''' </summary>
    Public ReadOnly Property ID As String
    ''' <summary>
    ''' The size of the drive
    ''' </summary>
    Public ReadOnly Property Size As String
    ''' <summary>
    ''' The model of the drive
    ''' </summary>
    Public ReadOnly Property Model As String
    ''' <summary>
    ''' The make of the drive
    ''' </summary>
    Public ReadOnly Property Vendor As String
    ''' <summary>
    ''' The speed of the drive
    ''' </summary>
    Public ReadOnly Property Speed As String
    ''' <summary>
    ''' The serial number
    ''' </summary>
    Public ReadOnly Property Serial As String
    ''' <summary>
    ''' List of partitions
    ''' </summary>
    Public ReadOnly Property Partitions As New Dictionary(Of String, Partition)

    ''' <summary>
    ''' Installs specified values parsed by Inxi to the class
    ''' </summary>
    Friend Sub New(ID As String, Size As String, Model As String, Vendor As String, Speed As String, Serial As String, Partitions As Dictionary(Of String, Partition))
        Me.ID = ID
        Me.Size = Size
        Me.Model = Model
        Me.Vendor = Vendor
        Name = $"{Vendor} {Model}"
        Me.Speed = Speed
        Me.Serial = Serial
        Me.Partitions = Partitions
    End Sub

End Class

File: Inxi.NET/Hardware/HardDrive.cs
using System;
using System.Collections.Generic;
using System.Linq;
using InxiFrontend.Base;
using InxiFrontend.Core;
using Newtonsoft.Json;

namespace InxiFrontend
{
    /// <summary>
    /// Hard drive class
    /// </summary>
    public sealed class HardDrive : IHardware, IEquatable<HardDrive>
    {
        [JsonProperty()]
        /// <summary>
        /// The name of hard drive
        /// </summary>
        public string Name { get; set; }
        [JsonProperty()]
        /// <summary>
        /// The udev ID of hard drive
        /// </summary>
        public string ID { get; private set; }
        [JsonProperty()]
        /// <summary>
        /// The size of the drive
        /// </summary>
        public string Size { get; private set; }
        [JsonProperty()]
        /// <summary>
        /// The model of the drive
        /// </summary>
        public string Model { get; private set; }
        [JsonProperty()]
        /// <summary>
        /// The make of the drive
        /// </summary>
        public string Vendor { get; private set; }
        [JsonProperty()]
        /// <summary>
        /// The speed of the drive
        /// </summary>
        public string Speed { get; private set; }
        [JsonProperty()]
        /// <summary>
        /// The serial number
        /// </summary>
        public string Serial { get; private set; }
        [JsonProperty()]
        /// <summary>
        /// List of partitions
        /// </summary>
        public Dictionary<string, Partition> Partitions { get; private set; }
        
        /// <summary>
        /// Installs specified values parsed by Inxi to the class
        /// </summary>
        internal HardDrive(string ID, string Size, string Model, string Vendor, string Speed, string Serial, Dictionary<string, Partition> Partitions)
        {
            this.ID = ID;
            this.Size = Size;
            this.Model = Model;
            this.Vendor = Vendor;
            Name = $"{Vendor} {Model}";
            this.Speed = Speed;
            this.Serial = Serial;
            this.Partitions = Partitions;
        }
        [JsonConstructor()]
        public HardDrive()
        {
        }

        public bool Equals(HardDrive other)
        {
            return HelperFunctions.AreObjectsEqual<HardDrive>(this, other, (x) => x.CustomAttributes.Any(y => y.AttributeType == typeof(JsonPropertyAttribute)) && x.Name != "Name");
        }

        public override bool Equals(object obj)
        {
            if (obj is not HardDrive)
            {
                return false;
            }
            return this.Equals((HardDrive)obj);
        }

        public override int GetHashCode()
        {
            return HelperFunctions.GetHashCodes(this, (x) => x.CustomAttributes.Any(y => y.AttributeType == typeof(JsonPropertyAttribute)));
        }
    }
}